### PR TITLE
[Doc] Add linters documentation

### PR DESCRIPTION
**Pull request by amenasria**
### What does this PR do?

This PR adds some documentation about the linters we're using on the datadog-agent repository.

### Motivation

Better developer experience.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided. Except if the `qa/skip-qa` label, with required either `qa/done` or `qa/no-code-change` labels, are applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by KevinFairise2 — APPROVED** [review]

File: docs/dev/linters.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
+# Linting the datadog-agent repository
+
+Welcome to our linters documentation ! Linters are tools that check our source code for errors, bugs, and stylistic inconsistencies. They're essential for maintaining code quality and consistency.
```
Comment (buraizu) [inline docs/dev/linters.md:3]:
```suggestion
Welcome to the Datadog Agent linters documentation! Linters are tools that check our source code for errors, bugs, and stylistic inconsistencies. They're essential for maintaining code quality and consistency.
```

File: docs/dev/linters.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+<!-- Those linters are running in the CI as required checks. -->
+
+Here are the linters we're using on the datadog-agent repository depending on the language:
```
Comment (buraizu) [inline docs/dev/linters.md:7]:
```suggestion
Here are the linters used on the datadog-agent repository depending on the language:
```

File: docs/dev/linters.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+## Go
+
+For Go we're using [golangci-lint](https://golangci-lint.run/), a Go linters aggregator. Its configuration is defined [in the .golangci.yml file](https://github.com/DataDog/datadog-agent/blob/main/.golangci.yml).
+
+The `linters` key define the list of linters we're using:
```
Comment (buraizu) [inline docs/dev/linters.md:13]:
```suggestion
The `linters` key defines the list of linters we're using:
```

File: docs/dev/linters.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+https://github.com/DataDog/datadog-agent/blob/dffd3262934a5540b9bf8e4bd3a743732637ef37/.golangci.yml#L65-L79
+
+To run the linters locally run `inv lint-go`.
```
Comment (buraizu) [inline docs/dev/linters.md:16]:
```suggestion
To run the linters locally, run `inv lint-go`.
```

File: docs/dev/linters.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+> [!TIP]
+> In your code, you can ignore a lint on line by prepending it with [the nolint directive](https://golangci-lint.run/usage/false-positives/#nolint-directive) e.g  `//nolint:linter_name`.
```
Comment (buraizu) [inline docs/dev/linters.md:19]:
```suggestion
> In your code, you can ignore linter issues on a line by prepending it with [the nolint directive](https://golangci-lint.run/usage/false-positives/#nolint-directive), for example,  `//nolint:linter_name`.
```

File: docs/dev/linters.md [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+> Example [here](https://github.com/DataDog/datadog-agent/blob/dffd3262934a5540b9bf8e4bd3a743732637ef37/cmd/agent/common/import.go/#L252) and [here](https://github.com/DataDog/datadog-agent/blob/dffd3262934a5540b9bf8e4bd3a743732637ef37/cmd/agent/common/misconfig/global.go/#L27-L32).
+
+
+## Python
+
+For Python we're using ([see invoke task](https://github.com/DataDog/datadog-agent/blob/dffd3262934a5540b9bf8e4bd3a743732637ef37/tasks/linter_tasks.py/#L17-L33)):
+- [flake8](https://flake8.pycqa.org/en/latest) a style linter.
+- [black](https://black.readthedocs.io/en/stable/) a code formatter.
+- [isort](https://pycqa.github.io/isort/) to sort the imports.
+- [vulture](https://github.com/jendrikseipp/vulture) to find unused code.
```
Comment (buraizu) [inline docs/dev/linters.md:29]:
```suggestion
- [flake8](https://flake8.pycqa.org/en/latest), a style linter.
- [black](https://black.readthedocs.io/en/stable/), a code formatter.
- [isort](https://pycqa.github.io/isort/), to sort the imports.
- [vulture](https://github.com/jendrikseipp/vulture), to find unused code.
```

File: docs/dev/linters.md [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+Their configuration is defined in both the [setup.cfg](https://github.com/DataDog/datadog-agent/blob/dffd3262934a5540b9bf8e4bd3a743732637ef37/setup.cfg) and the [pyproject.toml](https://github.com/DataDog/datadog-agent/blob/dffd3262934a5540b9bf8e4bd3a743732637ef37/pyproject.toml) files.
+
+To run the linters locally run `inv lint-python`.
```
Comment (buraizu) [inline docs/dev/linters.md:33]:
```suggestion
To run the linters locally, run `inv lint-python`.
```

File: docs/dev/linters.md [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+> [!TIP]
+> In your code, you can ignore a lint on line by prepending it with `# noqa: error_code`.
```
Comment (buraizu) [inline docs/dev/linters.md:36]:
```suggestion
> In your code, you can ignore linter issues on a line by prepending it with `# noqa: error_code`.
```

File: docs/dev/linters.md [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+> Example [here](https://github.com/DataDog/datadog-agent/blob/dffd3262934a5540b9bf8e4bd3a743732637ef37/tasks/new_e2e_tests.py/#L40-L42) and [here](https://github.com/DataDog/datadog-agent/blob/dffd3262934a5540b9bf8e4bd3a743732637ef37/tasks/release.py/#L257).
+
+## Troubleshooting
+
+### Go Q&A
```
Comment (buraizu) [inline docs/dev/linters.md:41]:
```suggestion
### Go
```

File: docs/dev/linters.md [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+Q: I get a lot of errors locally that I don't get in the CI, why ?
+
+A: This could come from several reasons:
```
Comment (buraizu) [inline docs/dev/linters.md:45]:
```suggestion
A: This could have several causes:
```

File: docs/dev/linters.md [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+- Your tools versions are not aligned with ours:
```
Comment (buraizu) [inline docs/dev/linters.md:46]:
```suggestion
- Your tool versions are not aligned with ours:
```

File: docs/dev/linters.md [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+    - `go version` should output the same as [the repository .go-version](https://github.com/DataDog/datadog-agent/blob/dffd3262934a5540b9bf8e4bd3a743732637ef37/.go-version).
+    - `golangci-lint --version` should output the same as [the repository internal/tools/go.mod](https://github.com/DataDog/datadog-agent/blob/dffd3262934a5540b9bf8e4bd3a743732637ef37/internal/tools/go.mod/#L8).
+- You're testing OS specific code, running locally the linters will only get the results for your local OS.
```
Comment (buraizu) [inline docs/dev/linters.md:49]:
```suggestion
- You're testing OS specific code; running locally, the linters only get the results for your local OS.
```

File: docs/dev/linters.md [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+- You didn't run `inv tidy-all` in the repository making some dependencies in the remote env outdated compare to your local env.
```
Comment (buraizu) [inline docs/dev/linters.md:50]:
```suggestion
- You didn't run `inv tidy-all` in the repository, making some dependencies in the remote env outdated compared to your local env.
```

File: docs/dev/linters.md [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+
+### About the new-from-rev golangci-lint parameter
+
+At some point we wanted to introduce the `revive` linter in th codebase. It was burdensome because enabling it made hundreds of errors appear in the CI. [We hence started using the `new-from-rev` parameter](https://github.com/DataDog/datadog-agent/blob/fcb19ce078e7969d285565beec5d374c5fd623e1/.golangci.yml#L65-L68) ([see doc](https://golangci-lint.run/usage/faq/#how-to-integrate-golangci-lint-into-large-project-with-thousands-of-issues)) to only display lints from changes made after commit `f40667d3841c6339be0d00d53e54a4a63f43f11e` (i.e when we enabled back `revive`).
```
Comment (buraizu) [inline docs/dev/linters.md:54]:
```suggestion
Introducing the `revive` linter in the codebase caused hundreds of errors to appear in the CI. As such, [the `new-from-rev` parameter](https://github.com/DataDog/datadog-agent/blob/fcb19ce078e7969d285565beec5d374c5fd623e1/.golangci.yml#L65-L68) was added to only display linter issues from changes made after the commit that enabled `revive`. [See the Golang documentation](https://golangci-lint.run/usage/faq/#how-to-integrate-golangci-lint-into-large-project-with-thousands-of-issues) for more information.
```

File: docs/dev/linters.md [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+
+We quickly realized [the debt that such a feature](https://github.com/golangci/golangci-lint/issues/4349) introduced:
+
+Let's say I have a legacy file hello.go with 100 lints (i.e linter errors). The new-from-rev parameter will silence them all. But if I rename this file to hello_world.go or move it to another folder then all the lints will rise again.
```
Comment (buraizu) [inline docs/dev/linters.md:58]:
```suggestion
In a scenario where you have a legacy file hello.go with 100 linter issues, the new-from-rev parameter removes them all. But if you rename the file to hello_world.go, or move it to another folder, all the linter issues reappear. See [issue 4349](https://github.com/golangci/golangci-lint/issues/4349) in the golangci repo for more information.
```

File: docs/dev/linters.md [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+
+Welcome to our linters documentation ! Linters are tools that check our source code for errors, bugs, and stylistic inconsistencies. They're essential for maintaining code quality and consistency.
+
+<!-- Those linters are running in the CI as required checks. -->
+
+Here are the linters we're using on the datadog-agent repository depending on the language:
+
+## Go
+
+For Go we're using [golangci-lint](https://golangci-lint.run/), a Go linters aggregator. Its configuration is defined [in the .golangci.yml file](https://github.com/DataDog/datadog-agent/blob/main/.golangci.yml).
```
Comment (buraizu) [inline docs/dev/linters.md:11]:
```suggestion
For Go, we're using [golangci-lint](https://golangci-lint.run/), a Go linters aggregator. Its configuration is defined [in the .golangci.yml file](https://github.com/DataDog/datadog-agent/blob/main/.golangci.yml).
```

File: docs/dev/linters.md [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+To run the linters locally run `inv lint-go`.
+
+> [!TIP]
+> In your code, you can ignore a lint on line by prepending it with [the nolint directive](https://golangci-lint.run/usage/false-positives/#nolint-directive) e.g  `//nolint:linter_name`.
+> Example [here](https://github.com/DataDog/datadog-agent/blob/dffd3262934a5540b9bf8e4bd3a743732637ef37/cmd/agent/common/import.go/#L252) and [here](https://github.com/DataDog/datadog-agent/blob/dffd3262934a5540b9bf8e4bd3a743732637ef37/cmd/agent/common/misconfig/global.go/#L27-L32).
+
+
+## Python
+
+For Python we're using ([see invoke task](https://github.com/DataDog/datadog-agent/blob/dffd3262934a5540b9bf8e4bd3a743732637ef37/tasks/linter_tasks.py/#L17-L33)):
```
Comment (buraizu) [inline docs/dev/linters.md:25]:
```suggestion
For Python, we're using ([see invoke task](https://github.com/DataDog/datadog-agent/blob/dffd3262934a5540b9bf8e4bd3a743732637ef37/tasks/linter_tasks.py/#L17-L33)):
```

**Review by buraizu — CHANGES_REQUESTED** [review]
Thanks for the PR, I left some mostly minor suggestions, but also reordered the last section a bit for clarity. Let me know any questions or feedback!

**Review by buraizu — APPROVED** [review]

**Pull request merged**